src/main.rs

```diff
@@ -69,29 +73,83 @@ impl Client {
     // Method should return an error when a connection already exists.
     // The client should send a handshake to the server.
     fn open(&mut self, addr: &str, server: Server) -> CommsResult<()> {
-        todo!()
+        match self.connections.get(addr) {
+            Some(_) => Err(CommsError::ConnectionExists(String::from(addr))),
+            None => {
+                self.connections
+                    .insert(String::from(addr), Connection::Open(server));
+                match self.send(
+                    addr,
+                    Message {
+                        msg_type: MessageType::Handshake,
+                        load: self.ip.clone(),
+                    },
+                ) {
+                    Err(_) => Err(CommsError::UnexpectedHandshake(String::from(addr))),
+                    Ok(_) => Ok(()),
+                }
+            }
+        }
     }
 
     // Sends the provided message to the server at the given `addr`.
     // Can only send messages through open connections. If the server
     // responds with a ServerLimitReached error, its corresponding connection
     // should be closed.
     fn send(&mut self, addr: &str, msg: Message) -> CommsResult<Response> {
-        // server.receive(msg)
-        todo!()
+        match msg.msg_type {
+            MessageType::Handshake => match self.connections.get_mut(addr) {
+                Some(Connection::Open(serv)) => match serv.receive(msg) {
+                    Err(CommsError::UnexpectedHandshake(addr2)) => {
+                        self.connections
+                            .insert(String::from(addr), Connection::Closed);
+                        Err(CommsError::UnexpectedHandshake(addr2))
+                    }
+                    _ => Ok(Response::HandshakeReceived),
+                },
+                _ => Err(CommsError::ConnectionNotFound(String::from(addr))),
+            },
+            MessageType::Post => match self.connections.get_mut(addr) {
+                Some(Connection::Open(serv)) => match serv.receive(msg) {
+                    Err(CommsError::ServerLimitReached(addr2)) => {
+                        self.connections
+                            .insert(String::from(addr), Connection::Closed);
+                        Err(CommsError::ServerLimitReached(addr2))
+                    }
+                    _ => Ok(Response::PostReceived),
+                },
+                Some(Connection::Closed) => Err(CommsError::ConnectionClosed(String::from(addr))),
+                None => Err(CommsError::ConnectionNotFound(String::from(addr))),
+            },
+            MessageType::GetCount => match self.connections.get_mut(addr) {
+                Some(Connection::Open(serv)) => serv.receive(msg),
+                Some(Connection::Closed) => Err(CommsError::ConnectionClosed(String::from(addr))),
+                None => Err(CommsError::ConnectionNotFound(String::from(addr))),
+            },
+        }
```

Comment:
Cała metoda jest niepotrzebnie skomplikowana - pattern matching jest w złej kolejności. `UnexpectedHandshake` oraz `ServerLimitReached` to błędy, które powinny być zgłaszane przez serwer i tutaj tylko przekazywane dalej.